### PR TITLE
fix for permutation circuit

### LINKED ISSUE #5812: bug in permutation circuits
The permutation circuits from the circuit library do not implement some permutation correctly. For example:

```py
from qiskit.circuit.library import Permutation
Permutation(8, [5,4,1,7,6,3,0,2]).decompose().draw()
```

here the 5th qubit has to end up in the 0th location, but it ends up in the 6th location because two swaps act on it.

```
                 
q_0: ────X─────X─
         │     │ 
q_1: ─X──┼──X──┼─
      │  │  │  │ 
q_2: ─┼──┼──X──┼─
      │  │     │ 
q_3: ─┼──┼──X──┼─
      │  │  │  │ 
q_4: ─X──┼──┼──┼─
         │  │  │ 
q_5: ────X──┼──┼─
            │  │ 
q_6: ───────┼──X─
            │    
q_7: ───────X────
                 
```

### PR DESCRIPTION
**Pull request by ewinston**
### Summary
fixes #5812 

### Details and comments
Not sure this is optimal solution but seems to work.

### CODE REVIEW

File: test/python/circuit/library/test_permutation.py [hunk 1]
```diff
@@ -38,6 +40,19 @@ def test_permutation_bad(self):
...
+        np.random.seed(1)
+        for length in range(5, 20, 2):
+            for _ in range(5):
+                permutation = np.random.permutation(length)
+                swap_list = _get_ordered_swap(permutation)
+                output = list(range(length))
+                for i, j in swap_list:
+                    output[i], output[j] = output[j], output[i]
+                self.assertTrue(np.array_equal(permutation, output))
+                self.assertLess(len(swap_list), length)
```
Comment (ajavadia) [inline test/python/circuit/library/test_permutation.py:54]: is it possible to write a functional test here that checks the Operator implemented by this circuit is equal to the expected Operator of a Permutation? The unitary associated with a permutation can be obtained by permuting columns of a 2^n x 2^n Identity matrix (though I haven't worked out how exactly). This particular test seems to rely on `_get_ordered_swap` being correct, which seems like an implementation detail.

File: qiskit/circuit/library/generalized_gates/permutation.py [hunk 2]
```diff
@@ -75,9 +75,19 @@ def __init__(self,
...
+
+def _get_ordered_swap(permutation_in):
+    permutation = list(permutation_in[:])
+    swap_list = []
+    for i, val in enumerate(permutation):
+        if val != i:
+            j = permutation.index(i)
+            swap_list.append((i, j))
+            permutation[i], permutation[j] = permutation[j], permutation[i]
+    return swap_list[::-1]
```
Comment (ajavadia) [inline qiskit/circuit/library/generalized_gates/permutation.py:93]: can you explain what this is doing (how the swaps are added)?
my thinking was that this problem could be approached as a sorting problem so doing something like Quicksort should give the minimal number of swaps to get the permutation.

File: test/python/circuit/library/test_permutation.py [hunk 3]
```diff
@@ -38,6 +40,19 @@ def test_permutation_bad(self):
         """Test that [0,..,n-1] permutation is required (no -1 for last element)."""
         self.assertRaises(CircuitError, Permutation, 4, [1, 0, -1, 2])
 
+    def test_get_ordered_swap(self):
+        """Test get_ordered_swap function."""
+        np.random.seed(1)
+        for length in range(5, 20, 2):
+            for _ in range(5):
```
[the remark this answers is not in the document]
Comment (ewinston) [reply test/python/circuit/library/test_permutation.py:47]: ok

File: qiskit/circuit/library/generalized_gates/permutation.py (hunk 2 above)
Comment (ewinston) [reply qiskit/circuit/library/generalized_gates/permutation.py:93]: I'll put a description into the doc string. As far as efficiency I compared this method to another based on "cycle decomposition" which is supposed to be efficient and this was as efficient. Here is a comparison trying all permutations of a given width for `_get_ordered_swap`, cycle decomposition and [a quick sort implementation here](https://stackabuse.com/quicksort-in-python/). It seems sometimes quick sort does more swaps.

| `(width, method)`  |   `max(num_swaps)` |   `sum(swaps)` |
|:--------------------|-----------------------:|-----------------------:|
| (5, 'cycle_decomp') |                      4 |          326           |
| (5, 'ordered_swap') |                      4 |          326           |
| (5, 'quick_sort')   |                      4 |          328           |
| (6, 'cycle_decomp') |                      5 |         2556           |
| (6, 'ordered_swap') |                      5 |         2556           |
| (6, 'quick_sort')   |                      6 |         2620           |
| (7, 'cycle_decomp') |                      6 |        22212           |
| (7, 'ordered_swap') |                      6 |        22212           |
| (7, 'quick_sort')   |                      7 |        23096           |
| (8, 'cycle_decomp') |                      7 |       212976           |
| (8, 'ordered_swap') |                      7 |       212976           |
| (8, 'quick_sort')   |                      9 |       225504           |
| (9, 'cycle_decomp') |                      8 |            2.23934e+06 |
| (9, 'ordered_swap') |                      8 |            2.23934e+06 |
| (9, 'quick_sort')   |                     11 |            2.40672e+06 |

File: test/python/circuit/library/test_permutation.py (hunk 1 above)
Comment (ewinston) [reply test/python/circuit/library/test_permutation.py:54]: ok

File: qiskit/circuit/library/generalized_gates/permutation.py (hunk 2 above)
Comment (ajavadia) [reply qiskit/circuit/library/generalized_gates/permutation.py:93]: nice, thanks for this detailed evaluation

**Review by ajavadia — COMMENTED** [review]
This looks good, can you just add a docstring for the method that finds the swap network, and add a releasenote.

**Review by kdk — APPROVED** [review]

**Pull request merged**